compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -116,7 +116,7 @@ object PatternMatcher {
     /** Widen type as far as necessary so that it does not refer to a pattern-
      *  generated variable.
      */
-    private def sanitize(tp: Type): Type = tp.widenExpr match {
+    private def sanitize(tp: Type): Type = tp.widenIfUnstable match {
```

Comment:
There is also `widenTermRefExpr` that seems to work here. Not sure which one is the most appropriate one. I defer to @odersky or @smarter

Reply:
Yeah the first commit uses `widenTermRefExpr`. Maybe @liufengyun can think of examples where widening singletons would make a difference?

Reply:
What about `x match { case _: x.type => ... }` ?

Reply:
We concluded with @smarter that there is no need to widen singleton types